src/Pact/Analyze/Translate.hs

```diff
@@ -1293,11 +1304,26 @@ translateNode astNode = withAstContext astNode $ case astNode of
             bindingTs
       recov <- view teRecoverability
       tid <- genTagId
-      emit $ TraceRequireGrant recov capName bindingTs $ Located (nodeInfo node) tid
-      pure $ Some SBool $ Enforce Nothing $ HasGrant tid cap vars
-
-  AST_ComposeCapability (AST_InlinedApp modName funName bindings appBodyA) ->
-    translateCapabilityApp modName (CapName $ T.unpack funName) bindings appBodyA
+      inScope <- Set.member capName <$> use tsStaticCapsInScope
+      if inScope then do
+        emit $ TraceRequireGrant recov capName bindingTs $ Located (nodeInfo node) tid
+        pure $ Some SBool $ Enforce Nothing $ HasGrant tid cap vars
+        else do
+        -- we have statically proven there is no way this capability can come into scope.
+        -- Weirdly, we want to warn about this instead of fail, as in many cases this is
+        -- intentional (functions "protected" from direct execution by a cap).
+        -- TODO consider some kind of declaration that would make this happen for all
+        -- properties of a defun. Right now properties are independent so there is
+        -- no way to "inform" properties of a protected function that a capability is
+        -- intentionally preventing direct execution.
```

Comment:
If we can prove that a particular capability can _never_ come into scope except through magical means, it might be useful to distinguish those as warnings so that unscoped caps could be errors again. Future Analysis work?

Reply:
My goal would be the opposite, the user would have to specify in the model that they _intended to protect the function_ with the cap. The problem is, this declaration has to be messaged to all the properties in a given function to avoid them being vacuous, so it probably requires a new term in the model language.

Reply:
Yeah, i imagined we'd have to track it in a term. Alright, future work 😄